### PR TITLE
User Endpoint

### PR DESCRIPTION
**Pull request by shatoboar**
Exposes endpoint under "/users/" that returns information on users.
Calls idp manager to get information not stored locally (email, name), which in the case of the managed version is auth0.

### CODE REVIEW

File: management/server/account.go [hunk 1]
```diff
@@ -214,9 +223,14 @@ func (am *DefaultAccountManager) GetAccountByUserOrAccountId(userId, accountId,
 	return nil, status.Errorf(codes.NotFound, "no valid user or account Id provided")
 }
 
+func isNil(i idp.Manager) bool {
```
Comment (mlsmaycon) [inline management/server/account.go:248]: Good catch. One thing, the function name is not matching the logic. You are returning true and it isn't nil.

File: management/server/account.go [hunk 2]
```diff
@@ -225,6 +239,53 @@ func (am *DefaultAccountManager) updateIDPMetadata(userId, accountID string) err
...
+	account, err := am.GetAccountById(accountID)
+	if err != nil {
+		return nil, err
+	}
+
+	queriedUsers := make([]*idp.UserData, 0)
+
+	// https://auth0.com/docs/manage-users/user-search/retrieve-users-with-get-users-endpoint#limitations
+	// auth0 limitation of 1000 users via this endpoint
+	for page := 0; page < 20; page++ {
```
Comment (mlsmaycon) [inline management/server/account.go]: This logic should belong to the idp manager for auth0

File: management/server/idp/auth0.go [hunk 3]
```diff
@@ -183,6 +186,126 @@ func (c *Auth0Credentials) Authenticate() (JWTToken, error) {
...
+	var list []*UserData
+	for user := range userData {
+		list = append(list, &userData[user])
+	}
+
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("unable to get UserData, statusCode %d", res.StatusCode)
+	}
+
+	return list, nil
```
Comment (mlsmaycon) [inline management/server/idp/auth0.go:268]: there should be a validation for an empty batch response, either here or at the API

File: management/server/idp/auth0.go [hunk 4]
```diff
@@ -183,6 +186,126 @@ func (c *Auth0Credentials) Authenticate() (JWTToken, error) {
...
+}
+
+// Requests user data from auth0
+// user data: email
+func (am *Auth0Manager) GetUserData(userId string, appMetadata AppMetadata) (*UserData, error) {
```
Comment (mlsmaycon) [inline management/server/idp/auth0.go]: It seems like you can use the same code and endpoint from GetBatchedUserData and only change the query parameter to use user id to retrieve the same result. @braginini what do you think?

**Review by mlsmaycon — CHANGES_REQUESTED** [review]
Besides tests for idp manager and account manager new methods, you can improve the logs as a lot of info logs you've added would be better placed as debug logs.

File: management/server/account.go [hunk 5]
```diff
@@ -225,6 +239,53 @@ func (am *DefaultAccountManager) updateIDPMetadata(userId, accountID string) err
...
+		batch, err := am.idpManager.GetBatchedUserData(accountID, page)
```
Comment (mlsmaycon) [inline management/server/account.go]: You should handle the case where there are no IDP manager credentials (self-hosted) where we usually don't call idp manager methods and in this case, `queriedUsers` will always be empty.

Comment (braginini) [reply management/server/account.go]: Good point @mlsmaycon 
I also think we need a test for this

File: management/server/account.go (hunk 2 above)
Comment (braginini) [reply management/server/account.go]: @mlsmaycon agree

File: management/server/account.go (hunk 1 above)
Comment (shatoboar) [reply management/server/account.go:248]: done.

File: management/server/idp/auth0.go [hunk 6]
```diff
@@ -183,6 +186,136 @@ func (c *Auth0Credentials) Authenticate() (JWTToken, error) {
...
+		log.Debugf("Requested batch; %v", batch)
+
+		defer func() {
+			err = res.Body.Close()
+			if err != nil {
+				log.Errorf("error while closing update user app metadata response body: %v", err)
+			}
+		}()
+
+		if len(batch) == 0 {
```
Comment (mlsmaycon) [inline management/server/idp/auth0.go]: this check should be performed after StatusCode as a non 200 status will cause len(batch) == 0

File: management/server/idp/auth0.go [hunk 7]
```diff
@@ -183,6 +186,136 @@ func (c *Auth0Credentials) Authenticate() (JWTToken, error) {
...
+			list = append(list, &batch[user])
+		}
+	}
+
+	return list, nil
+}
+
+// Requests user data from auth0
+// user data: email
+func (am *Auth0Manager) GetUserDataByEmail(userId string, appMetadata AppMetadata) (*UserData, error) {
```
Comment (mlsmaycon) [inline management/server/idp/auth0.go]: name here seems to be misleading as we use the user's id and not email

File: management/server/idp/auth0.go (hunk 6 above)
Comment (shatoboar) [reply management/server/idp/auth0.go]: Ah thanks, good point

**Review by braginini — APPROVED** [review]

**Pull request merged**